docs/proposals/block-storage.md

```diff
@@ -25,90 +25,89 @@ lower bound).
+Here's a list of requirements for a good solution:
 
-  * If you want to back the database up you would have to copy the blocks as
-    well, while in reality the blocks can always be redownloaded and we
-    don't care much about losing them.
+  * R1. Store blocks separately from the main database.
 
-  * A hard drive failure has a much bigger chance to corrupt the database,
-    if the database contains lots of data (blocks).
+  * R2. Fast lookup by header hash.
+  
+  * R3. Fast sequential iteration
 
-So, here's a list of requirements for a good solution:
+  * R4. A solution should be stable to corruptions
 
-  * We'd like to store blocks separately from the main database.
+  * R5. Apply compression to the blocks, and since each block isn't
+        very well compressible on its own, the solution should use either
+        dictionary compression or compress all blocks as a whole.
+    
+  * R6. Solution should be easy to implement and maintain
 
-  * We want fast lookup by header hash.
+## Possible solutions
 
-  * We'd like to store them in batches (e.g. one batch = one epoch), so that
-    corruption of one batch wouldn't affect other batches.
+### Storing all blocks in one RocksDB
 
-  * We want to apply compression to the blocks, and since each block isn't
-    very well compressible on its own, the solution should use either
-    dictionary compression or compress all blocks as a whole.
+  1. New RocksDB for blocks. Enable out of a box compression with dictionary.
+  
+  2. When a new block comes, add it along with undo to this database by header hash.
 
-  * It'd be good to be able to reuse somebody else's code, because generally
-    there are lots of non-obvious problems with files (e.g. whenever you're
-    modifying a file and want to protect against power failure, you should
-    use `flush` in the right way – and it's very easy to forget about that).
+Such solution satisfies R1, R2, R5 and R6.
+Also there is a way to achieve R3, it will be described in the next chapter.
 
-## The solution
-
-Out of all considered solutions (SQLite, storing blocks in an archive, CBOR
-map, some database) it seems that the ideal solution is very simple: 
+### One RocksDB per epoch
 
   1. Create a new RocksDB database for each epoch, with keys = hashes and
-     values = blocks. Enable compression with dictionary.
+     values = blocks. Enable out of a box compression with dictionary.
 
   2. Store a map `hash -> epoch` in some other database (e.g. “blockindex”).
 
-  3. When a new block comes, add it to the database and to the index map.
+  3. When a new block comes, add it along with undo to the database and to the index map.
+  
+This solution satisfies R1, R2, R4, R5.
+Also there is the same way to achieve R3 like for **Storing all blocks in one RocksDB**.
+  
+### One plain file per epoch
+
+  1. Create a new file for each epoch.
+  
+  2. Store epoch's blocks (along with undo) in plain file sequentially.
```

Comment:
Why undo? What is needed there exactly? Can't it be as simple as just the CBOR encoding of all blocks of the epoch (in a list)?